### PR TITLE
[Agent] Team 4 Outgoing custom client

### PR DESCRIPTION
**Pull request by MikeIsMike**
# Summary

Please provide a summary of the change.

## Additional Information

Auxiliary information goes here.

## Test Plan

- Add information on how you tested your changes.

### CODE REVIEW

File: internal/clients/team4/client.go [hunk 1]
```diff
@@ -2,16 +2,207 @@
...
+		clientJudge:   judge{BaseJudge: &baseclient.BaseJudge{}, t: nil},
+		clientSpeaker: speaker{BaseSpeaker: &baseclient.BaseSpeaker{}},
+		yes:           "",
+		obs:           &observation{},
+		internalParam: &internalParameters{},
+		savedHistory:  &map[uint]map[shared.ClientID]judgeHistoryInfo{},
+	}
+	team4client.clientJudge.parent = &team4client
+	team4client.clientSpeaker.parent = &team4client
+
```
Comment (lhl2617) [inline internal/clients/team4/client.go]: can you move this outside to a `NewClient` function? This will preserve your state between runs on the same instance, I'm afraid.

Comment (lhl2617) [reply internal/clients/team4/client.go]: see how team1 is doing it

File: internal/clients/team4/judge.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	"os"
+	"testing"
+
+	"github.com/SOMAS2020/SOMAS2020/internal/common/baseclient"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/rules"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+)
+
+func init() {
+	os.Remove("./Historymap.txt")
```
Comment (lhl2617) [inline internal/clients/team4/judge.go]: yo what?

File: internal/clients/team4/judge.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	return id
+}
+
+func (j *judge) logf(format string, a ...interface{}) {
+	if j.t != nil {
+		j.t.Log(fmt.Sprintf(format, a...))
+	}
+}
+
+func dump(filename string, format string, v ...interface{}) {
```
Comment (lhl2617) [inline internal/clients/team4/judge.go]: can delete?

File: internal/common/shared/iigoroles.go [hunk 4]
```diff
@@ -41,8 +41,9 @@ func (r Role) MarshalJSON() ([]byte, error) {
 // RuleVoteType provides enumerated values for Approving, Rejecting or Abstaining from a vote.
 type RuleVoteType int
 
+// type of vote for each ballot
 const (
-	Approve = iota
+	Approve RuleVoteType = iota
```
Comment (lhl2617) [inline internal/common/shared/iigoroles.go]: ah, thanks for this. can you submit this in a separate PR?

File: internal/common/rules/rulematrix.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ type RuleMatrix struct {
 func (r *RuleMatrix) RuleMatrixIsEmpty() bool {
 	if r.RuleName == "" &&
 		len(r.RequiredVariables) == 0 &&
-		r.Mutable == false &&
+		!r.Mutable &&
```
Comment (lhl2617) [inline internal/common/rules/rulematrix.go]: similar as below.

**Review by lhl2617 — CHANGES_REQUESTED** [review]
as per comments

File: internal/common/shared/iigoroles.go (hunk 4 above)
Comment (MikeIsMike) [reply internal/common/shared/iigoroles.go]: Check PR #372

File: internal/common/rules/rulematrix.go (hunk 5 above)
Comment (MikeIsMike) [reply internal/common/rules/rulematrix.go]: Check PR #372

File: internal/common/shared/iigoroles.go (hunk 4 above)
Comment (MikeIsMike) [reply internal/common/shared/iigoroles.go]: PR #372

**Comment by MikeIsMike** [comment]
> as per comments

Please merge it for me if all is ok. Thanks!

**Comment by MikeIsMike** [comment]
> as per comments

Please merge for me when all is ok.

File: internal/clients/team4/client.go [hunk 6]
```diff
@@ -2,16 +2,209 @@
...
+	//trustMatrix *mat.Dense //this shouldn't be in internal parameters
+	greediness    float64
+	selfishness   float64
+	fairness      float64
+	collaboration float64
+	riskTaking    float64
+	agentsTrust   []float64
+}
+
+type personality struct {
```
Comment (lhl2617) [inline internal/clients/team4/client.go:90]: unused?

File: internal/clients/team4/client.go [hunk 7]
```diff
@@ -2,16 +2,209 @@
...
+
+//Overriding the StartOfTurn method of the BaseClient
+func (c *client) StartOfTurn() {
+	c.yes = "yes"
+	c.Logf(`what are you doing?
+	=========================================
+	==============================================
+	================================================`)
+	c.Logf("this is a %v for you ", c.yes)
+	fmt.Println(reflect.TypeOf(c))
```
Comment (lhl2617) [inline internal/clients/team4/client.go:128]: still used? :P

File: internal/clients/team4/client.go [hunk 8]
```diff
@@ -2,16 +2,209 @@
...
+	for i := 0; i < ruleMatrix.AuxiliaryVector.Len(); i++ {
+		currentAuxValue := ruleMatrix.AuxiliaryVector.AtVec(i)
+		for j := range ruleMatrix.RequiredVariables {
+
+			idealValue := idealRuleMatrix.ApplicableMatrix.At(i, j)
+			actualValue := ruleMatrix.ApplicableMatrix.At(i, j)
+
+			if currentAuxValue == 0 {
+				// ==0 condition
+				distance += math.Abs(idealValue-actualValue) / idealValue
```
Comment (lhl2617) [inline internal/clients/team4/client.go:165]: div0 check please

File: internal/clients/team4/iigo.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+	// lyingLevel will be positive when agent is inclined to lie.
+	lyingLevel := mat.Dot(importance, parameters) - lyingThreshold
+
+	// Construct output struct.
+	resReported := currentResources
+
+	// Agent lies linearly based on lyingLevel.
+	if lyingLevel > 0 {
+		resReported = currentResources * (1 / shared.Resources((lyingLevel + 1)))
```
Comment (lhl2617) [inline internal/clients/team4/iigo.go:181]: div0 check

**Review by lhl2617 — APPROVED** [review]
pre-approving. please fix div0 checks.

File: internal/clients/team4/iigo.go (hunk 9 above)
Comment (MikeIsMike) [reply internal/clients/team4/iigo.go:181]: denom is strictly greater than 0 here, still need an explicit check?

Comment (lhl2617) [reply internal/clients/team4/iigo.go:181]: no need then :)
